webhooks/virtualmachine/mutation/virtualmachine_mutator_cdrom.go

```diff
@@ -43,286 +38,287 @@ func MutateCdromControllerOnUpdate(
 		ctrlID := pkgutil.ControllerID{
-			ControllerType: controllerType,
+			ControllerType: ctrlType,
 			BusNumber:      busNum,
 		}
-		if _, exists := usedSlotMap[ctrlID]; !exists {
-			if nextBusNum == nil {
-				nextBusNum = &busNum
-			}
-			continue
-		}
 
-		for unitNum := range maxSlotCount {
-			if !usedSlotMap[ctrlID][unitNum] {
-				return &ControllerSlot{
-					BusNumber:      busNum,
-					UnitNumber:     unitNum,
-					ControllerType: controllerType,
-				}, wasMutated
+		if _, exists := occupiedSlots[ctrlID]; exists {
+			unitNum := vmopv1util.NextAvailableUnitNumber(
+				vmopv1util.CreateNewController(ctrlType, busNum, ""),
+				occupiedSlots[ctrlID])
+			if unitNum != -1 {
+				return busNum
 			}
+		} else if firstAvailableBusNumber == -1 {
+			firstAvailableBusNumber = busNum
 		}
-
-	}
-
-	if nextBusNum != nil {
-
-		if addControllerAndInitSlotMap(
-			usedSlotMap, hwSpec, controllerType, *nextBusNum) {
-			wasMutated = true
-		}
-
-		return &ControllerSlot{
-			BusNumber:      *nextBusNum,
-			UnitNumber:     0,
-			ControllerType: controllerType,
-		}, wasMutated
 	}
 
-	return nil, wasMutated
+	return firstAvailableBusNumber
 }
 
-// constructUsedSlotMap builds a map of used controller slots from the VM's
-// hardware spec and status. It initializes slot maps for IDE and SATA
-// controllers and marks slots as used by non-CD-ROM devices to avoid
-// conflicts when assigning new CD-ROM devices.
-func constructUsedSlotMap(
-	vm *vmopv1.VirtualMachine) map[pkgutil.ControllerID][]bool {
+// constructOccupiedSlots builds a map of occupied controller slots from the
+// VM's hardware spec and status. It tracks which slots are occupied by
+// non-CD-ROM devices to avoid conflicts when assigning new CD-ROM devices.
+func constructOccupiedSlots(
+	vm *vmopv1.VirtualMachine) map[pkgutil.ControllerID]sets.Set[int32] {
 
-	usedSlotMap := make(map[pkgutil.ControllerID][]bool)
+	occupiedSlots := make(map[pkgutil.ControllerID]sets.Set[int32])
 
 	if vm.Spec.Hardware == nil {
-		return usedSlotMap
+		return occupiedSlots
 	}
 
+	// Initialize sets for each controller in the spec.
 	for _, controller := range vm.Spec.Hardware.IDEControllers {
-		usedSlotMap[pkgutil.ControllerID{
+		ctrlID := pkgutil.ControllerID{
 			ControllerType: vmopv1.VirtualControllerTypeIDE,
 			BusNumber:      controller.BusNumber,
-		}] = make([]bool, pkgutil.IDEControllerMaxSlotCount)
+		}
+		occupiedSlots[ctrlID] = sets.New[int32]()
 	}
 
 	for _, controller := range vm.Spec.Hardware.SATAControllers {
-		usedSlotMap[pkgutil.ControllerID{
+		ctrlID := pkgutil.ControllerID{
 			ControllerType: vmopv1.VirtualControllerTypeSATA,
 			BusNumber:      controller.BusNumber,
-		}] = make([]bool, pkgutil.SATAControllerMaxSlotCount)
+		}
+		occupiedSlots[ctrlID] = sets.New[int32]()
 	}
 
 	if vm.Status.Hardware == nil {
-		return usedSlotMap
+		return occupiedSlots
 	}
 
+	// Mark slots occupied by non-CD-ROM devices.
 	for _, ctrl := range vm.Status.Hardware.Controllers {
```

Comment:
Just to confirm, for CDROM, we have no choice but to check the vm.status for occupied slots for now

Reply:
We are checking occupied slots that are used by volumes here. I guess we could merge this into the hardware mutator in the future if we want to directly respect the vm.spec for occupied slots. 

I think we are doing the same thing in the hardware mutator (check occupied slots for cd-rom devices). 

I will create a JIRA to follow up on this.